### ISSUE TITLE
Early DMA check in ubuntu fails

### ISSUE THREAD

**Issue by philipandag** [Dasharo/dasharo-issues#985]
### Component

Dasharo firmware

### Device

NovaCustom V54 14th Gen

### Dasharo version

v0.9.1-rc1

### Test case ID

EDP001.001

### Brief summary

Early DMA protection check using cbmem fail

### How reproducible

100% in two retries

### How to reproduce

Do the EDP001.001 test manually
- enable early DMA protection
- boot to ubuntu
- run `sudo ./cbmem -1`

### Expected behavior

output should contain:
[DEBUG]  VT-d @ 0xfed91000, version 5.0
[INFO ]  Setting DMA protection [0x0 - 0x46c00000]
[INFO ]  Setting DMA protection [0x100000000 - 0x00000008afc00000]
[INFO ]  Successfully enabled VT-d PMR DMA protection

### Actual behavior

```
root@3mdeb:/home/ubuntu# ./cbmem -1 | grep -i "dma"
?ACPI:    * DMAR
?soc_fill_dmar - gfxvtbar:0xfc800000  0xfc800001
root@3mdeb:/home/ubuntu# 
```

```
root@3mdeb:/home/ubuntu# ./cbmem -1 | grep -i "protection"
?BM-LOCKDOWN: Skipping enabling boot media protection
?ME: SPI Protection Mode Enabled : NO
root@3mdeb:/home/ubuntu# 
```

**Comment by mkopec** [comment]
The code to configure DMA protection is not wired up on soc/intel/meteorlake

**Comment by mkopec** [comment]
Most likely just need to apply https://review.coreboot.org/c/coreboot/+/68450 but for meteorlake and that should be enough

**Comment by philipandag** [comment]
I am applying the patch here: https://github.com/Dasharo/coreboot/pull/553. We need the Intel FSP to compile and test it on the target platform though.

**Comment by philipandag** [comment]
After applying the patch `cbmem -l` shows
```
[ERROR]  VT-d PMR HOB not found, not enabling DMA protection
```

**Comment by philipandag** [comment]
The test passes on V560TNE with v0.9.1-rc4.

**Comment by philipandag** [comment]
Not working on V540TND with v0.9.1-rc5
```
ubuntu@3mdeb:~$ grep -i "vt-d" cbmem-dma-enabled.log
[DEBUG]  VT-d @ 0xfc801000, version 7.0
[ERROR]  VT-d PMR HOB not found, not enabling DMA protection
```

[cbmem-dma-enabled.log](https://github.com/user-attachments/files/16964278/cbmem-dma-enabled.log)
[cbmem-dma-disabled.log](https://github.com/user-attachments/files/16964276/cbmem-dma-disabled.log)

**Comment by mkopec** [comment]
Looks like the hob is generated on IoT FSP

**Comment by SebastianCzapla** [comment]
Issue still present in v0.9.1-rc6.
[cbmem-dma-enabled.txt](https://github.com/user-attachments/files/17289985/cbmem-dma-enabled.txt)

**Comment by SebastianCzapla** [comment]
Issue still present in v0.9.1-rc7.
[cbmem-dma-enabled.txt](https://github.com/user-attachments/files/17415580/cbmem-dma-enabled.txt)

**Comment by mkopec** [comment]
@SebastianCzapla The option was supposed to be hidden, how are you testing if it's not visible?

**Comment by SebastianCzapla** [comment]
You are right, it is not visible in the security options. I misread other option for it, my bad.

**Comment by mkopec** [comment]
We attempted to simply enable DMA protection in coreboot the same way it was done for alder lake, but it looks like the FSP does not produce the required HOB

**Comment by tlaurion** [comment]
Also affects nv4x on Heads master per dasahro coreboot commit used: 94e5f5d5b808cf8d8fd5c70d4ef6a08a054f8986
https://github.com/linuxboot/heads/blob/462c157b23d4e45b80a7685d289f9a6d14d53fdd/modules/coreboot#L94C1-L99C46

Note that nv41 coreboot config under master doesn't have early dma protection on:
https://github.com/linuxboot/heads/blob/462c157b23d4e45b80a7685d289f9a6d14d53fdd/config/coreboot-novacustom-nv4x_adl.config#L426

----

`cbmem -1` log captured on Heads recovery shell with Heads master:

[withhout_early_dma_protection-master.log](https://github.com/user-attachments/files/18943595/withhout_early_dma_protection-master.log)


When turning "CONFIG_ENABLE_EARLY_DMA_PROTECTION=y", `cbmem -1` log:
[with_early_boot_with_vtd_DISABLED_WARNING.log](https://github.com/user-attachments/files/18943607/with_early_boot_with_vtd_DISABLED_WARNING.log)

Exerpt: 
> [INFO ]  VT-d DMA protection disabled by option

**Comment by tlaurion** [comment]
Cross Ref https://github.com/linuxboot/heads/pull/1913#issuecomment-2678763157

**Comment by tlaurion** [comment]
Please 
- add nv41 as affected
- rename issue as "Early DMA protection cannot be applied to alderlake+" 
- add in release notes for latest coreboot+uefi/ coreboot+heads

CC @macpijan

**Comment by macpijan** [comment]
> Also affects nv4x on Heads master per Dasharo coreboot commit used:

Thanks for leaving a note here. We will verify this when working on the future NV4 coreboot relases.

The UEFI release v1.7.2 https://docs.dasharo.com/variants/novacustom_nv4x_adl/releases/#v172-2024-01-03 points specifically to https://github.com/Dasharo/dasharo-issues/issues/630 so I believe we have tested that it used to work correctly back then.

**Comment by tlaurion** [comment]
CC @macpijan

    Crossref to downstream user created issue https://github.com/linuxboot/heads/issues/1922

**Comment by tlaurion** [comment]
> Early DMA protection disabled for v560tu/v540tu at [Dasharo/coreboot#769](https://github.com/Dasharo/coreboot/pull/769)

Can we expect a fix since FSP used now was updated?

Still known issue over https://docs.dasharo.com/variants/novacustom_v540tu/releases_heads/#v091-2026-05-11
